[WIP] pine64-pinephonepro: Init

samueldr opened the pull request:
<img src="https://user-images.githubusercontent.com/132835/146710731-beacfbb4-0287-416f-bca2-7191af2abfd0.jpg" height="256">

* * *

### TODO

 - [ ] Try building a "normal" and modular NixOS-flavoured kernel
 - [ ] Write instructions

* * *

<details>
<summary>
Outdated instructions...
</summary>

### Quick install guide

**This will overwrite anything on the target eMMC and SD cards.**

> First, [*neuter* the pre-installed system on the eMMC of the Pinephone Pro](https://wiki.pine64.org/wiki/PinePhone_Pro_Developer_Edition#Nuking_the_factory_AOSP_installation).

This will use the temporary installer script, here's the command to build the "hello" example system.

```
 $ nix-build --argstr device pine64-pinephonepro -A outputs.temp-tow-boot-install-script examples/hello/
```

 1. Build Tow-Boot from https://github.com/Tow-Boot/Tow-Boot/pull/67
    ```
     $ nix-build -A pine64-pinephonePro
    ```
 2. Write Tow-Boot to an SD card (via e.g. `dd`)
    ```
    sudo dd if=result/shared.disk-image.img of=/dev/sdXXX bs=8M oflag=direct,sync
    ```
 3. Insert the SD card in the Pinephone Pro, and boot it while holding *Volume Up*, making sure your *Pinephone Pro* is connected via USB to your computer. The LED will turn blue when in *USB Mass Storage* mode.
 4. Write Tow-Boot to the eMMC (via e.g. `dd`)
    ```
     $ sudo dd if=result/shared.disk-image.img of=/dev/sdXXX bs=8M oflag=direct,sync
    ```
 5. Remove the SD card. We will make sure the installation to eMMC is successful.
 6. Reboot the Pinephone Pro by holding power, press *Volume Up* when it's booting back up so it boots again in *USB Mass Storage* mode.
 7. "Install" to the eMMC via the installer script:
    ```
     $ sudo /nix/store/[...]-temp-tow-boot-install/install.sh /dev/sdXXX
    ```

Once the script is done, you can reboot the device by holding the power button. You don't need to hold any of the volume keys to continue booting in the Mobile NixOS system.

#### Updating the boot partition

Simple!

 1. Boot the device to *USB Mass Storage* by holding *Volume Up*.
 2. Write (`dd`) the boot partition to what should be the fourth partition in a "normal" setup.

### NAQ (never asked questions)

#### Why's this so damned hard?

Actually this is not much harder than "just" using `dd` to write an SD card with, let's say, Target Disk Mode, and then "just" using `dd` to write to the eMMC now attached via USB.

The main added *weirdness* is managing the platform firmware separately.

This is done to [better exercise planning for a better NixOS on ARM](https://discourse.nixos.org/t/planning-for-a-better-nixos-on-arm/15346). If it can work on "complex" devices like a phone, surely it'll be trivial on a "boring" SBC!

Additionally, some of the steps outlined here are **universal** for any distribution that will not distribute the *platform firmware* bundled into their system images.

Finally, this is a *temporary measure*.

The permanent solution will be an *installer* system. The user will write Tow-Boot to the eMMC, and boot from a universal SD card that holds an installer system that will guide the user through some questions about partitioning (e.g. ask about FDE and filesystems), the preferred graphical environment, and then do a "classic" opinionated 
`nixos-install` under the covers.



</details>

samueldr committed:
hardware-rockchip: Add rockchip-rk3399s

samueldr committed:
pine64-pinephonepro: Init kernel

Config is the defconfig from mobian, arbitrarily changed.

samueldr committed:
kernel-config: Enable RD_GZIP/XZ

samueldr committed:
pine64-pinephonepro: Add firmware

This is based on the pinebook pro firmware package. Unknown of this is
all of the required firmware, and if there's too much.

samueldr committed:
pine64-pinephonepro: Add USB role switch task

This is used to force the type-c direction.

samueldr committed:
pine64-pinephonepro: Init

samueldr committed:
================================================================================

samueldr committed:
================================================================================

samueldr committed:
pine64-pinephonepro: Enable default LED behaviour

samueldr committed:
pine64-pinephonepro: WIP kernel config validation

samueldr committed:
pine64-pinephonepro: WIP shell on stage-1 fail

samueldr commented on modules/hardware-rockchip.nix:
```diff
@@ -12,6 +11,11 @@ in
+    hardware.socs.rockchip-rk3399s.enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = "enable when SOC is RK3399";
```
```suggestion
      description = "enable when SOC is RK3399s";
```

samueldr commented on modules/kernel-config.nix:
```diff
@@ -36,6 +36,9 @@ in
         TMPFS_POSIX_ACL = yes;
         TMPFS_XATTR = yes;
 
+        RD_GZIP = yes;
+        RD_XZ = yes;
+
```
Move this change (and related normalization) into a new PR.

telent commented:
I believe that
 ```
 $ nix-build --argstr device pine64-pinephonepro -A outputs.temp-tow-boot-install-script examples/hello/
```
in these instructions should be 

``` 
$ nix-build --argstr device pine64-pinephonepro -A outputs.u-boot.temp-tow-boot-install-script examples/hello/
```

(there's a `u-boot` in the attribute path)

samueldr commented:
The *system type*'s outputs are copied into the main `outputs` attrset.

https://github.com/NixOS/mobile-nixos/blob/796040d35e38507e8a9ed97c760cba6e923a8ca9/lib/eval-with-configuration.nix#L48-L49

So while you're right that the attribute exists under the `outputs.u-boot` namespace, it also exists at the root level. This is because of backward compatibility.

tomfitzhenry commented:
<img src="https://user-images.githubusercontent.com/61303/148962372-aa5a1dcd-d922-4bec-b044-2fb6fcad51ba.jpeg" width="400">

This works for me, great!

I like the idea and execution of Tow-Boot.

iclanzan commented on devices/pine64-pinephonepro/default.nix:
```diff
@@ -0,0 +1,98 @@
+    soc = "rockchip-rk3399s";
+    ram = 1024 * 4;
+    screen = {
+      width = 1920; height = 1080;
```
```suggestion
      width = 720; height = 1440;
```

samueldr committed:
kernel-config: Set FW_LOADER_USER_HELPER_FALLBACK=n

wentam committed:
pine64-pinephonepro: audio support

wentam committed:
pine64-pinephonepro: Update kernel to 5.16.0

And change origin to the now decided location of the community
maintained kernel.

wentam committed:
pine64-pinephonepro: vibrator support

samueldr committed:
pine64-pinephonepro: kernel 5.16.0 → 5.16.7

samueldr committed:
pine64-pinephonepro: Use appropriate frequency scaling

See https://gitlab.com/pine64-org/linux/-/merge_requests/36

TLDR: chip too hot, designed to run cooler slower.

samueldr committed:
kernel-config: Use HZ_1000

samueldr committed:
kernel-config: Restart on panic

samueldr committed:
pine64-pinephonepro: ensure SPI config is enabled

samueldr committed:
pine64-pinephonepro: Normalize kernel config

samueldr commented:
Thanks you @wentam for some of the changes.

Rebased, dropped needless changes from the history.

wentam commented:
Perhaps we should run [this patch](https://gitlab.com/pine64-org/linux/-/merge_requests/34) as well? I've been running that patch for a couple of days and it makes the charging situation **much** better.

samueldr commented:
> Perhaps we should run [this patch](https://gitlab.com/pine64-org/linux/-/merge_requests/34) as well? I've been running that patch for a couple of days and it makes the charging situation **much** better.

Probably!

samueldr committed:
pine64-pinephonepro: Add charging workarounds

See:

 - https://gitlab.com/pine64-org/linux/-/merge_requests/34

samueldr committed:
[HACK] pine64-pinephonepro: Add workaround to show the logo

This serves more to prove to me that it actually should be working. What
needs to actually be done is figure out *what* is preventing the logo
from showing.

Note that this is not a Pinephone Pro issue, nor is it an RK3399 issue,
nor is it an ARM issue. This workaround had been authored for an x86_64
system initially.

samueldr commented:
Two changes:

 - Added the suggested change set for charging, and I'm observing that indeed it won't get stuck in a state where it's stating it's charging, but actually drawing 0A (from a USB power monitor).
 - Added a workaround for what I consider a bug/regression in mainline Linux and `CONFIG_LOGO`

* * *

What's that about `CONFIG_LOGO`? I've found that upstream seems intent on making sure the kernel logo is never used anymore at all. I don't know where in the stack it happens, but display init is being deferred in a way that is detrimental to user experience *in my opinion*.

I assume the goal is that, with blinders on, you'll see it makes sense to defer changing the display mode until needed as it prevents needless flickers during boot, e.g. your normal UEFI x86_64 laptop will continue showing its splash until some userspace component draws to the display.

Here this is not ideal since there is no display init done by the platform firmware. Having some sort of feedback early on that things are happening is good.

Though *even if we had feedback from the platform firmware*, I would argue that being able to control this behaviour would be beneficial here as we would want to show it ASAP to actually show "yes the kernel is booting"; otherwise if it crashes really early it will look like the platform firmware is hanging, when in reality it's happening somewhere between kernel init and userspace init.
